### PR TITLE
New instrument driver for Anaheim Automation stepper motor controllers.

### PR DESCRIPTION
**Pull request by samcondon4**
Just another instrument basic instrument driver.

### CODE REVIEW

File: pymeasure/instruments/anaheimautomation/stepmotorcontroller.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+
+    def slew(self, direction):
+        """Sends the slew command to the motor controller. This tells the controller to clock the stepper motor until a stop command is sent or a limit switch is reached.
+        
+        :param direction: value to set on the direction property before sending the slew command to the controller.
+        """
+        self.direction = direction
+        self.write("S") 
+
+    async def async_step(self, steps, direction, **kwargs):
```
[the remark this answers is not in the document]
Comment (samcondon4) [reply pymeasure/instruments/anaheimautomation/stepmotorcontroller.py:200]: Thank you for providing this example. This seems to achieve what I want. My usecase is in a telescope focus calibration measurement where we are using a whole slew of stepper motors (driven by these anaheim automation controllers) to set the incidence angle of light onto our IR detectors. I am developing a gui interface to control and automate this measurement. During a measurement where we are testing the detector response at multiple incidence angles I need to be able to pend until all motors arrive in their final position before starting an exposure but I do not want to block the operation of the gui and control of other instruments while this pend occurs. I have been using asyncio in early versions of this interface, but your point about the red/blue function problem is well taken.

So with this concurrent.futures implementation, are you thinking that this should be a part of the driver itself or just given as an example of how concurrency can be achieved?

File: pymeasure/instruments/anaheimautomation/stepmotorcontroller.py [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+#
+
+import asyncio
+from pyvisa.errors import VisaIOError
+from pymeasure.adapters import VISAAdapter
+from pymeasure.instruments import Instrument
+from pymeasure.instruments.validators import truncated_range, strict_discrete_set
+
+
+class MotorController(Instrument):
```
[the remark this answers is not in the document]
Comment (samcondon4) [reply pymeasure/instruments/anaheimautomation/stepmotorcontroller.py:32]: I have used both the DPY50601 and DPE25601 stepper controllers from the product group that you link. I know that at least for these two products the command set is identical. I just glanced over a few other controllers (DPC50501, DPN10601, and DPD75601) and the command sets are nearly identical here as well. The DPC50501 has fewer features than the other controllers so the command set is only a subset of those available to some of the other controllers. However, as far as I can tell, the features that I have currently implemented are universal within the product group.

Perhaps we name this class something generic like DPController and consider it a base-class containing the commands that are identical between these controllers and for cases where some controllers have additional features these can be implemented in subclasses? Or would you rather keep additional base-classes out of this layer of the instruments package?

[the remark this answers is not in the document]
Comment (samcondon4) [reply pymeasure/instruments/anaheimautomation/stepmotorcontroller.py]: @bilderbuchi 
Let me know what you think of the new file/class names.

[the remark this answers is not in the document]
Comment (bilderbuchi) [reply pymeasure/instruments/anaheimautomation/stepmotorcontroller.py]: > Perhaps we name this class something generic like DPController and consider it a base-class containing the commands that are identical between these controllers and for cases where some controllers have additional features these can be implemented in subclasses

This is the way! :-D

File: pymeasure/instruments/anaheimautomation/stepmotorcontroller.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (bilderbuchi) [reply pymeasure/instruments/anaheimautomation/stepmotorcontroller.py:200]: Awesome application! :-)
> I do not want to block the operation of the gui and control of other instruments while this pend occurs.

Absolutely! I think this _the_ classical case of the GUI and business logic being separated and happening in separate threads. E.g. when using pyqt this is pretty standard. Don't know your specifics, but i found [this book](https://www.packtpub.com/product/mastering-gui-programming-with-python/9781789612905) to be quite approachable.

The way I see it, the concurrency stuff should not be part of the driver, but sit in the application layer, i.e. between your GUI code and the instrument drivers. 

In your case I could see you configuring a move with GUI element, then having a button "Start slew!". When pressed, it kicks of the concurrent move(s). Futures have a [way of notifying](https://docs.python.org/3.9/library/concurrent.futures.html#concurrent.futures.Future.add_done_callback) via callback when they're done, you can probably use that to then re-activate the button or make an indicator green or unlock the "Expose" button or whatnot - most GUI frameworks revolve around callbacks (signals/slots in Qt lingo).

Good seeing! :telescope:

File: pymeasure/instruments/anaheimautomation/stepmotorcontroller.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        """
+        return self.adapter.values("@{}".format(self.idn) + command + self.write_termination, **kwargs)
+        
+    def ask(self, command):
+            """ Override the instrument base ask method to add the motor controller's id to the command string.
+            
+            :param command: command string to be sent to the instrument
+            """
+            return self.adapter.ask("@{}".format(self.idn) + command + self.write_termination)
+
```
[the remark this answers is not in the document]
Comment (samcondon4) [reply pymeasure/instruments/anaheimautomation/stepmotorcontroller.py]: Hi, this device has no reset command as it performs a reset automatically on power up. There are also no commands to close communication with the instrument as it is always just listening over its serial port.

File: pymeasure/instruments/anaheimautomation/stepmotorcontroller.py [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+        "Reads the current value of the error codes register."
+    )
+
+    def __init__(self, resourceName, idn, encoder=False, **kwargs):
+        """
+        Initialize communication with the motor controller with id=idn. In addition to the keyword arguments that can be
+        set for the Instrument base class, this class has the following kwargs:
+
+        :param idn: (int) Assigned id of the motor controller.
+        :param encoder: (bool) Flag to indicate if an encoder has been connected to the controller.
```
[the remark this answers is not in the document]
Comment (samcondon4) [reply pymeasure/instruments/anaheimautomation/stepmotorcontroller.py]: Agreed

File: pymeasure/instruments/anaheimautomation/stepmotorcontroller.py [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+        return self._encoder_enabled
+
+    @encoder_enabled.setter
+    def encoder_enabled(self, en):
+        """ Set the value of the _encoder_enabled flag. When set, asynchronous coroutine methods like step_async() will query
+            the "encoder_position" property instead of the "position" property.
+
+        :param en: (bool) boolean value to set the _encoder_enabled flag with
+        """
+        if type(en) is bool:
```
[the remark this answers is not in the document]
Comment (samcondon4) [reply pymeasure/instruments/anaheimautomation/stepmotorcontroller.py]: Not particularly concerned about this. Changed it to your suggestion.

File: pymeasure/instruments/anaheimautomation/stepmotorcontroller.py [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+        """Method that sends the G command to the controller to tell the controller to turn the 
+           motor the number of steps previously set with the steps property."""
+
+        self.write("G")
+
+    def step(self, steps, direction):
+        """Similar to the go() method, but also sets the number of steps and the direction in the same method.
+        
+        :param steps: Number of steps to clock
+        :param direction: Value to set on the direction property before sending the go command to the controller.
```
[the remark this answers is not in the document]
Comment (samcondon4) [reply pymeasure/instruments/anaheimautomation/stepmotorcontroller.py]: Done

File: pymeasure/instruments/anaheimautomation/stepmotorcontroller.py [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+        """
+        self.steps = steps 
+        self.direction = direction
+        self.write("G")
+
+    def slew(self, direction):
+        """Sends the slew command to the motor controller. This tells the controller to clock the stepper motor until a stop command is sent or a limit switch is reached.
```
[the remark this answers is not in the document]
Comment (samcondon4) [reply pymeasure/instruments/anaheimautomation/stepmotorcontroller.py]: "move the stepper motor" is better here. Changed this.

File: pymeasure/instruments/anaheimautomation/stepmotorcontroller.py [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+        self.step(steps, direction)
+        # wait until the step operation completes. # 
+        await self.wait_for_completion(**kwargs)
+
+    def write(self, command):
+        """Override the instrument base write method to add the motor controller's id to the command string.
+        
+        :param command: command string to be sent to the motor controller.
+        """
+        self.adapter.write("@{}".format(self.idn) + command)
```
[the remark this answers is not in the document]
Comment (samcondon4) [reply pymeasure/instruments/anaheimautomation/stepmotorcontroller.py]: Got it.

File: pymeasure/instruments/anaheimautomation/stepmotorcontroller.py [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+       """An integer property that represents the step position as seen by the motor controller. This property can be set.""",
+       validator=truncated_range,
+       values=[-8388607, 8388607],
+    ) 
+
+    encoder_position = Instrument.measurement(
+        "VEP", 
+        """An integer property that represents the step position as counted by an externally connected encoder. This property cannot be set, but can be 
+        reset to 0 using the reset_encoder_position() method.""",
+    )
```
[the remark this answers is not in the document]
Comment (samcondon4) [reply pymeasure/instruments/anaheimautomation/stepmotorcontroller.py]: I like this suggestion, though one issue I see with merging these two properties is that the `motor position` can be set to any value within the range that I have specified in the current `position` property, while the `encoder position` can only be reset to 0. I think mixing these properties could make this fact less transparent. What are your thoughts on only allowing the position to be reset to 0? So there would be a single `position` property to query either the motor or encoder position as you have described, then a separate method `reset_position()` is created that always resets both the motor position and the encoder position.

[the remark this answers is not in the document]
Comment (samcondon4) [reply pymeasure/instruments/anaheimautomation/stepmotorcontroller.py]: For my use case, I will never need to be able to set the `motor position` to an arbitrary value. I will only ever need to reset it to 0 when performing homing operations. My suspicion is that this will be the case for the vast majority of users as well.

File: pymeasure/instruments/anaheimautomation/stepmotorcontroller.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (samcondon4) [reply pymeasure/instruments/anaheimautomation/stepmotorcontroller.py]: Thank you for the book suggestion! I will definitely give it a look. I went ahead and removed the async methods. I will plan to submit a new PR with an example use of the concurrency feature within my application layer once I have this implemented!

File: pymeasure/instruments/anaheimautomation/stepmotorcontroller.py [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+    def ask(self, command):
+        """ Override the instrument base ask method to add the motor controller's id to the command string.
+
+        :param command: command string to be sent to the instrument
+        """
+        return self.adapter.ask("@{}".format(self.idn) + command)
+   
+    async def wait_for_completion(self, interval=0.5, threshold=3):
+        """ Query the motor controller's step (or encoder) position and only return when inactivity is detected. Inactivity is 
+            defined as when the previous queried position and current queried position match for "threshold" number of times.
```
[the remark this answers is not in the document]
Comment (samcondon4) [reply pymeasure/instruments/anaheimautomation/stepmotorcontroller.py]: Upon further inspection of the command set, the controller does in fact have a "busy" query which allows one to directly check if the controller is moving a motor. This query would be definitely be better suited in this case. I went ahead and removed this function since it is an async function, but added a `busy` property to perform this query anyway.

File: pymeasure/instruments/anaheimautomation/stepmotorcontroller.py [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+       "VZ", "Z%i",
+       """An integer property that represents the step position reference as seen by the motor controller. This property can be set.""",
+       validator=truncated_range,
+       values=[-8388607, 8388607]
+    ) 
+
+    error_reg = Instrument.measurement(
+        "!",
+        "Reads the current value of the error codes register."
+    )
```
[the remark this answers is not in the document]
Comment (samcondon4) [reply pymeasure/instruments/anaheimautomation/stepmotorcontroller.py]: @LongnoseRob Hi, I've added an IntFlag subclass to decode error codes read by the error_reg property, though I am not completely sure how the logging should be achieved. Can you point me in the direction of existing instruments that do this?

[the remark this answers is not in the document]
Comment (LongnoseRob) [reply pymeasure/instruments/anaheimautomation/stepmotorcontroller.py]: @samcondon4 have a look [here](https://github.com/pymeasure/pymeasure/blob/122f0e7a4269e4dfce6de7e8deebdb920367e65b/pymeasure/instruments/hp/hp3478A.py#L357-L369)

File: pymeasure/instruments/anaheimautomation/stepmotorcontroller.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (bilderbuchi) [reply pymeasure/instruments/anaheimautomation/stepmotorcontroller.py]: >  So there would be a single position property to query either the motor or encoder position as you have described, then a separate method reset_position() is created that always resets both the motor position and the encoder position.

Sounds good. Maybe call it `home()` or something similar if that matches what you do.

> For my use case, I will never need to be able to set the motor position to an arbitrary value. I will only ever need to reset it to 0 when performing homing operations. My suspicion is that this will be the case for the vast majority of users as well.

Hm, it seems I have misunderstood something. I was under the impression that you can also command the motor to move by setting a position, i.e. `motor.position = 1234` moves to that point (e.g. by using  `step()`in the background. If that is not the case, why? Would that not be a quite natural interface?

File: pymeasure/instruments/anaheimautomation/stepmotorcontroller.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (bilderbuchi) [reply pymeasure/instruments/anaheimautomation/stepmotorcontroller.py]: Also, naming it `check_errors()` [our default name for this](https://pymeasure.readthedocs.io/en/latest/api/instruments/instruments.html#pymeasure.instruments.Instrument.check_errors) enables your properties to use it if they set the `get|set_check_errors` flag.

File: pymeasure/instruments/anaheimautomation/stepmotorcontroller.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (samcondon4) [reply pymeasure/instruments/anaheimautomation/stepmotorcontroller.py]: I can certainly move the driver in that direction. Originally, the `position` property was only used to get/set the value of the controller's internal position reference. It wouldn't actually move the motor at all. Though I think being able to do `motor.position = 1234` and have the motor move to that position would be quite nice. I'd also like to distinguish between _step position_ and _absolute position_. The motor controller always counts position measured in steps, though in my use case I'd like to be able to command an _absolute_ position in say meters, inches, etc. 

A potential way this can be accomplished is to have two properties, one called `step_position` which behaves exactly like the current `position` property but with the added feature that you can run `motor.step_position = <desired position>` and the motor will actually move to that position. The other property would be called `absolute_position` and it also behaves exactly the same way, but calls unit conversion methods `absolute_to_steps()` when setting a new position and `steps_to_absolute()` when querying position. These methods would of course have to be implemented in a subclass as the mapping between steps and absolute position will vary depending on what the user is driving with the controller. What do you think of this type of interface? My next commit will include my idea for this implementation.

[the remark this answers is not in the document]
Comment (samcondon4) [reply pymeasure/instruments/anaheimautomation/stepmotorcontroller.py]: > Sounds good. Maybe call it `home()` or something similar if that matches what you do.
The problem with calling it home() is that homing is a command that can be sent to the controller, so I'd like to have a separate home() method to perform this operation. I've just implemented this method and kept the name `reset_position()` for the other method.

File: pymeasure/instruments/anaheimautomation/dpseriesmotorcontroller.py [hunk 12]
```diff
@@ -0,0 +1,377 @@
...
+        """An integer property that represents the allowable error in encoder pulses from the desired
+           position before the encoder auto-correct function runs. This property can be set.""",
+        validator=truncated_range,
+        values=[0, 255],
+    )
+
+    busy = Instrument.measurement(
+            "VF",
+            """Query to see if the controller is currently moving a motor."""
+    )
```
Comment (bilderbuchi) [inline pymeasure/instruments/anaheimautomation/dpseriesmotorcontroller.py:164]:
```suggestion
    busy = Instrument.measurement(
        "VF",
        """Query to see if the controller is currently moving a motor."""
    )
```

File: pymeasure/instruments/anaheimautomation/dpseriesmotorcontroller.py [hunk 13]
```diff
@@ -0,0 +1,377 @@
...
+        :return error_code: one byte with the error codes register contents
+        """
+        current_errors = self.error_reg
+        if current_errors != 0:
+            logging.error("DP-Series motor controller error detected: %s" % current_errors)
+        return current_errors
+    
+    def __init__(self, resourceName, address=0, encoder_enabled=False, **kwargs):
+        """
+        Initialize communication with the motor controller with id=idn. In addition to the keyword arguments that can be
```
Comment (bilderbuchi) [inline pymeasure/instruments/anaheimautomation/dpseriesmotorcontroller.py:184]: What's this "id=idn" that is mentioned here? Left-over from a previous iteration?

File: pymeasure/instruments/anaheimautomation/dpseriesmotorcontroller.py [hunk 14]
```diff
@@ -0,0 +1,377 @@
...
+    def encoder_enabled(self):
+        """
+        Return the value of the _encoder_enabled flag.
+        """
+        return self._encoder_enabled
+
+    @encoder_enabled.setter
+    def encoder_enabled(self, en):
+        """ Set the value of the _encoder_enabled flag. When set, asynchronous coroutine methods like step_async() will
+            query the "encoder_position" property instead of the "position" property.
```
Comment (bilderbuchi) [inline pymeasure/instruments/anaheimautomation/dpseriesmotorcontroller.py:216]: Docstring references to async stuff that is not there anymore.

File: pymeasure/instruments/anaheimautomation/dpseriesmotorcontroller.py [hunk 15]
```diff
@@ -0,0 +1,377 @@
...
+        """A string property that represents the direction in which the stepper motor will rotate upon subsequent 
+           step commands. This property can be set. 'CW' corresponds to clockwise rotation and 'CCW' corresponds to 
+           counter-clockwise rotation.""",
+        map_values=True, 
+        validator=strict_discrete_set,
+        values={"CW" : "+", "CCW" : "-"},
+        get_process=lambda d: "+" if d == 1.0 else "-",
+    )    
+
+    steps = Instrument.control(
```
Comment (bilderbuchi) [inline pymeasure/instruments/anaheimautomation/dpseriesmotorcontroller.py:112]: Maybe a better name to help disambiguate between the different variations of steps/step/position: `planned_steps`?

File: pymeasure/instruments/anaheimautomation/dpseriesmotorcontroller.py [hunk 16]
```diff
@@ -0,0 +1,377 @@
...
+        else:
+            pos = self.ask("VZ")    
+        return int(pos)
+    
+    @step_position.setter
+    def step_position(self, pos):
+        """ Sends command to the motor controller to move to the desired step position. Note that in DP series
+            controller instrument manuals, this property corresponds to the 'absolute position' command. In this driver,
+            the `absolute_position` property implements a conversion between steps to absolute units whereas the DP
+            series manuals refer to `absolute position` as measured in steps.
```
Comment (bilderbuchi) [inline pymeasure/instruments/anaheimautomation/dpseriesmotorcontroller.py:239]: great that you explain this in detail, for clarity! Watch out, though, Python properties only get one docstring, I think it's the one from the getter? Best combine them into one. You can check with `myproperty._doc__`, or in IPython `myproperty?`.

File: pymeasure/instruments/anaheimautomation/dpseriesmotorcontroller.py [hunk 17]
```diff
@@ -0,0 +1,377 @@
...
+        
+        :param abs_pos: absolute position in units determined by the absolute_to_steps() method.
+        """
+        steps_pos = self.absolute_to_steps(abs_pos)
+        self.step_position = steps_pos
+    
+    def absolute_to_steps(self, pos):
+        """ Convert an absolute position to a number of steps to move. This must be implemented in subclasses.
+        
+        :param pos: Absolute position in the units determined by the subclassed position_to_steps() method.
```
Comment (bilderbuchi) [inline pymeasure/instruments/anaheimautomation/dpseriesmotorcontroller.py:270]: Still old method name, from a renaming I guess.

File: pymeasure/instruments/anaheimautomation/dpseriesmotorcontroller.py [hunk 18]
```diff
@@ -0,0 +1,377 @@
...
+    def stop(self):
+        """Method that stops all motion on the motor controller."""
+        self.write(".")
+
+    def go(self):
+        """Method that sends the G command to the controller to tell the controller to turn the 
+           motor the number of steps previously set with the steps property."""
+        self.write("G")
+
+    def step(self, steps, direction):
```
Comment (bilderbuchi) [inline pymeasure/instruments/anaheimautomation/dpseriesmotorcontroller.py:298]: Might be worth it to name this one `step_distance` or `step_for()` to increase clarity.

File: pymeasure/instruments/anaheimautomation/dpseriesmotorcontroller.py [hunk 19]
```diff
@@ -0,0 +1,377 @@
...
+        # check if an address related command was sent. #
+        if "%" in command or "~" in command:
+            vals = super().values("@%s" % command, **kwargs)
+        else:
+            vals = super().values("@%i%s" % (self.address, command))
+
+        return vals
+    
+    def ask(self, command):
+        """ Override the instrument base ask method to add the motor controller's id to the command string.
```
Comment (bilderbuchi) [inline pymeasure/instruments/anaheimautomation/dpseriesmotorcontroller.py:358]: At first I was wondering why you'd redo this, then I realised that this surfaces an interesting inconsistency in our API - the VISAAdapter is the only one using a dedicated method for `ask` (PyVISA's `query`), the others just reuse their own `read()` and `write()` -- in those cases I fear you'll prepend the id twice. (You can maybe test with a `SerialAdapter`).
As a workaround, I think you should be able to check in your overridden `write` for `command.startswith("@")`, which means the command has already been preprocessed by `ask`.

**Review by bilderbuchi — CHANGES_REQUESTED** [review]
This is nearly finished I think, apart from some further comments/requests.

Take the following as a comment only, by someone who previously fought with another stepper motor controller with a 400 page manual (:D), not(!) a request to remodel. I fully understand if you consider it "good enough" and just want to get this in.

In general, I feel a bit confused by the naming and API: slewing and moving relatively are methods, positions are a property; there is `slew` and `go`, which sounds pretty similar effect-wise; the sometimes slim distinction between step-the-verb (as in the method of that name) and step-the-attribute (in various args) and "step-the-noun" (in the property `steps`) vs. "position" vs `step_position`;
I think it could be clearer to have a triplet of methods like `move` (slew), `move_for` (steps, in direction), `move_to` (a position), but I understand why you want a `position` property to query. maybe add a thin wrapper method for the latter? 
Also, should `step` and `go` be `_step` and `_go`, i.e. "private" properties, to reduce the API surface a bit? Or do you envision yourself using this a lot, e.g. in preprogrammed but delayed moves?

File: pymeasure/instruments/anaheimautomation/stepmotorcontroller.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (bilderbuchi) [reply pymeasure/instruments/anaheimautomation/stepmotorcontroller.py]: sounds good re: the examples!

File: pymeasure/instruments/anaheimautomation/dpseriesmotorcontroller.py (hunk 17 above)
Comment (bilderbuchi) [reply pymeasure/instruments/anaheimautomation/dpseriesmotorcontroller.py:270]: otherwise this approach of punting implementation to a subclass is fine by me, it will always depend on the physical setup by the user anyway, and this way they can implement more complex transformations than a simple multiplication-by-gear-ratio, e.g. if the motor drives a cam or whatnot.

File: pymeasure/instruments/anaheimautomation/dpseriesmotorcontroller.py (hunk 13 above)
Comment (samcondon4) [reply pymeasure/instruments/anaheimautomation/dpseriesmotorcontroller.py:184]: Thanks for catching this. Yes, this is leftover from a previous iteration.

**Comment by samcondon4** [comment]
> In general, I feel a bit confused by the naming and API: slewing and moving relatively are methods, positions are a property; there is `slew` and `go`, which sounds pretty similar effect-wise; the sometimes slim distinction between step-the-verb (as in the method of that name) and step-the-attribute (in various args) and "step-the-noun" (in the property `steps`) vs. "position" vs `step_position`; I think it could be clearer to have a triplet of methods like `move` (slew), `move_for` (steps, in direction), `move_to` (a position), but I understand why you want a `position` property to query. maybe add a thin wrapper method for the latter? Also, should `step` and `go` be `_step` and `_go`, i.e. "private" properties, to reduce the API surface a bit? Or do you envision yourself using this a lot, e.g. in preprogrammed but delayed moves?

Very good point. It is quite easy to whip up an api that works without thinking about how the language may make it more difficult for others to use. In general, I do not like going with solutions that are just "good enough" so I appreciate your scrutiny. Here are my thoughts on addressing this:

- Remove `go()`: This method seems unnecessary as I do not envision myself running preprogrammed but delayed moves, and I think it just clutters up the api.
- Change `slew()` to `move()`: This will behave exactly the same as before, but `move()` seems to make for more clear terminology.
- Remove the current `steps` property. I think this is more clutter in the api for the same reason as the `go()` method.
- Keep the `step_position` and `absolute_position` properties unchanged. 
- Remove the current `step()` method entirely. This could be changed to a method called `move_for()` but the same effect can be accomplished by running for example, `step_position += 1234`

In summary, the above changes leave the api with just four methods/properties to move and stop a motor, those being `step_position`, `absolute_position`, `move()`, and of course `stop()`. 

I will commit these changes in the next iteration. Let me know what you think.

File: pymeasure/instruments/anaheimautomation/dpseriesmotorcontroller.py (hunk 19 above)
Comment (samcondon4) [reply pymeasure/instruments/anaheimautomation/dpseriesmotorcontroller.py]: Good catch. Yeah when using the SerialAdapter the id was prepended twice. I added a check in write so that if @ is seen in the command string, then it just directly sends the command string as is.

File: pymeasure/instruments/anaheimautomation/dpseriesmotorcontroller.py [hunk 20]
```diff
@@ -0,0 +1,357 @@
...
+    """Base class to interface with Anaheim Automation DP series stepper motor controllers. This driver has been tested
+       with the DPY50601 and DPE25601 motor controllers.
+    """
+
+    address = Instrument.control(
+        "%", "~%i",
+        """Integer property representing the address that the motor controller uses for serial communications.""",
+        validator=strict_range,
+        values=[0, 99],
+        get_process=lambda addr: int(addr),
```
Comment (bilderbuchi) [inline pymeasure/instruments/anaheimautomation/dpseriesmotorcontroller.py:84]: Did you see [this comment](https://github.com/samcondon4/pymeasure/commit/78eaee15d1a310eb53476c424dc23b894a484eb3#r60793080)?

> note: this is not well-documented at the moment, but if you pass `cast=int` here instead this does the same thing. (leftover kwargs get passed on to `adapter.values()`)

*btw, `int` is a callable just like a lambda, so there's no need to define a lambda here anyway.*

cf. #530

File: pymeasure/instruments/anaheimautomation/dpseriesmotorcontroller.py [hunk 21]
```diff
@@ -0,0 +1,357 @@
...
+        return int(pos)
+    
+    @step_position.setter
+    def step_position(self, pos):
+        step_pos = int(pos)
+        if -8388607 < step_pos < 8388607:
+            self.write("P%i" % step_pos)
+            self.write("G")
+        else:
+            raise ValueError("Provided step position out of valid range!")
```
Comment (bilderbuchi) [inline pymeasure/instruments/anaheimautomation/dpseriesmotorcontroller.py:239]: you can use the built-in `validators.strict_range` here:
```
step_pos = strict_range(pos, (-8388607, 8388607))
self.write("P%i" % step_pos)
self.write("G")
```
(the int() cast is also done by the string formatter %i)

**Review by bilderbuchi — CHANGES_REQUESTED** [review]
Awesome, thank you for your understanding. I like it! 
2 small improvements, then we're done I think.

**Comment by samcondon4** [comment]
Great, thanks for the suggestions! This driver has improved significantly from the first commit.

**Comment by bilderbuchi** [comment]
Thank you for your contribution! :-) :tada:
